dbms/src/Common/assert_cast.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <type_traits>
+#include <typeinfo>
+#include <typeindex>
+#include <string>
+
+#include <Common/Exception.h>
+#include <common/demangle.h>
+
+
+namespace DB
+{
+    namespace ErrorCodes
+    {
+        extern const int BAD_CAST;
+    }
+}
+
+
+/** Perform static_cast in release build.
+  * Checks type by comparing typeid and throw an exception in debug build.
+  * The exact match of the type is checked. That is, cast to the ancestor will be unsuccessful.
+  */
+template <typename To, typename From>
+To assert_cast(From && from)
+{
+#ifndef NDEBUG
+    try
+    {
+        if constexpr (std::is_pointer_v<To>)
+        {
+            if (typeid(*from) == typeid(std::remove_pointer_t<To>))
```

Comment:
We should compare typeid().hash_code:

_There is no guarantee that the same std::type_info instance will be referred to by all evaluations of the typeid expression on the same type, although std::type_info::hash_code of those type_info objects would be identical, as would be their std::type_index._

-- from https://en.cppreference.com/w/cpp/language/typeid

Reply:
Yes, it can be different across DSOs. We can also improve it for `typeid_cast` but need to check for performance.